### PR TITLE
Add script hooks for typed/modular items, refactor futuristic items

### PR DESCRIPTION
**Pull request by Ada18980**
Refactors the futuristic functions, putting them in `Screens/Inventory/Futuristic/Futuristic.js`, and converts the following items to Typed/Modular:

Futuristic Crate is now Modular
VR Headset is now Modular
Interactive Visor is now Typed
Futuristic Mask is now Typed
Futuristic Heels are now Typed
Futuristic Shoes are now Typed
Futuristic Leg Cuffs are now Typed
Futuristic Cuffs are now Typed
Futuristic Ankle Cuffs are now Typed
Futuristic Armbinder is now Typed
Futuristic Straitjacket is now Modular
Futuristic Earphones are now Typed
Futuristic Mittens are now Typed
Futuristic Bra is now Modular

### CODE REVIEW

File: BondageClub/Tools/Node/AssetCheck.js [hunk 1]
```diff
@@ -7,10 +7,11 @@ const BASE_PATH = "../../";
 const neededFiles = [
 	"Scripts/Common.js",
 	"Scripts/Dialog.js",
+	"Screens/Inventory/Futuristic/Futuristic.js",
 	"Scripts/ModularItem.js",
 	"Scripts/TypedItem.js",
 	"Assets/Female3DCG/Female3DCG.js",
-	"Assets/Female3DCG/Female3DCGExtended.js"
+	"Assets/Female3DCG/Female3DCGExtended.js",
```
Comment (Jomshir98) [inline BondageClub/Tools/Node/AssetCheck.js:14]:
```suggestion
	"Assets/Female3DCG/Female3DCGExtended.js"
```

File: BondageClub/Scripts/ModularItem.js [hunk 2]
```diff
@@ -101,7 +102,7 @@ function ModularItemRegister(asset, config) {
  */
 function ModularItemCreateLoadFunction(data) {
 	const loadFunctionName = `${data.functionPrefix}Load`;
-	window[loadFunctionName] = function () {
+	let originalFunction = function () {
```
Comment (Jomshir98) [inline BondageClub/Scripts/ModularItem.js:105]:
```suggestion
	const loadFunction = function () {
```

File: BondageClub/Scripts/ModularItem.js [hunk 3]
```diff
@@ -111,6 +112,11 @@ function ModularItemCreateLoadFunction(data) {
 		}
 		DialogExtendedMessage = DialogFindPlayer(`${data.dialogSelectPrefix}${ModularItemBase}`);
 	};
+	if (data.scriptHooks && data.scriptHooks.load) {
+		window[loadFunctionName] = function () {
+			data.scriptHooks.load(originalFunction);
+		};
+	} else window[loadFunctionName] = originalFunction;
```
Comment (Jomshir98) [inline BondageClub/Scripts/ModularItem.js:119]:
```suggestion
	if (data.scriptHooks && data.scriptHooks.load) {
		window[loadFunctionName] = function () {
			data.scriptHooks.load(loadFunction);
		};
	} else window[loadFunctionName] = loadFunction;
```

File: BondageClub/Scripts/ModularItem.js [hunk 4]
```diff
@@ -120,11 +126,16 @@ function ModularItemCreateLoadFunction(data) {
  */
 function ModularItemCreateDrawFunction(data) {
 	const drawFunctionName = `${data.functionPrefix}Draw`;
-	window[drawFunctionName] = function () {
+	let originalFunction = function () {
```
Comment (Jomshir98) [inline BondageClub/Scripts/ModularItem.js:129]:
```suggestion
	const drawFunction = function () {
```

File: BondageClub/Scripts/ModularItem.js [hunk 5]
```diff
@@ -120,11 +126,16 @@ function ModularItemCreateLoadFunction(data) {
...
 		const currentModule = data.currentModule || ModularItemBase;
 		const drawFunction = data.drawFunctions[currentModule];
 		return drawFunction();
 	};
+	if (data.scriptHooks && data.scriptHooks.draw) {
+		window[drawFunctionName] = function () {
+			data.scriptHooks.draw(originalFunction);
+		};
+	} else window[drawFunctionName] = originalFunction;
```
Comment (Jomshir98) [inline BondageClub/Scripts/ModularItem.js:138]:
```suggestion
	if (data.scriptHooks && data.scriptHooks.draw) {
		window[drawFunctionName] = function () {
			data.scriptHooks.draw(drawFunction);
		};
	} else window[drawFunctionName] = drawFunction;
```

File: BondageClub/Scripts/ModularItem.js [hunk 6]
```diff
@@ -134,11 +145,30 @@ function ModularItemCreateDrawFunction(data) {
  */
 function ModularItemCreateClickFunction(data) {
 	const clickFunctionName = `${data.functionPrefix}Click`;
-	window[clickFunctionName] = function () {
+	let originalFunction = function () {
```
Comment (Jomshir98) [inline BondageClub/Scripts/ModularItem.js:148]:
```suggestion
	const clickFunction = function () {
```

File: BondageClub/Scripts/ModularItem.js [hunk 7]
```diff
@@ -134,11 +145,30 @@ function ModularItemCreateDrawFunction(data) {
...
 		const currentModule = data.currentModule || ModularItemBase;
 		const clickFunction = data.clickFunctions[currentModule];
 		return clickFunction();
 	};
+	if (data.scriptHooks && data.scriptHooks.click) {
+		window[clickFunctionName] = function () {
+			data.scriptHooks.click(originalFunction);
+		};
+	} else window[clickFunctionName] = originalFunction;
```
Comment (Jomshir98) [inline BondageClub/Scripts/ModularItem.js:157]:
```suggestion
	if (data.scriptHooks && data.scriptHooks.click) {
		window[clickFunctionName] = function () {
			data.scriptHooks.click(clickFunction);
		};
	} else window[clickFunctionName] = clickFunction;
```

File: BondageClub/Scripts/ModularItem.js [hunk 8]
```diff
@@ -147,7 +177,7 @@ function ModularItemCreateClickFunction(data) {
  * @param {ModularItemConfig} config - The item's extended item configuration
  * @returns {ModularItemData} - The generated modular item data for the asset
  */
-function ModularItemCreateModularData(asset, { Modules, ChatSetting, ChangeWhenLocked }) {
+function ModularItemCreateModularData(asset, { Modules, ChatSetting, ChangeWhenLocked, ScriptHooks}) {
```
Comment (Jomshir98) [inline BondageClub/Scripts/ModularItem.js:180]:
```suggestion
function ModularItemCreateModularData(asset, { Modules, ChatSetting, ChangeWhenLocked, ScriptHooks }) {
```

File: BondageClub/Scripts/TypedItem.js [hunk 9]
```diff
@@ -71,7 +72,7 @@ function TypedItemRegister(asset, config) {
  * @returns {TypedItemData} - The generated typed item data for the asset
  */
 function TypedItemCreateTypedItemData(asset,
-	{ Options, Dialog, ChatTags, Dictionary, ChatSetting, DrawImages, ChangeWhenLocked, Validate }
+	{ Options, Dialog, ChatTags, Dictionary, ChatSetting, DrawImages, ChangeWhenLocked, Validate, ScriptHooks}
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:75]:
```suggestion
	{ Options, Dialog, ChatTags, Dictionary, ChatSetting, DrawImages, ChangeWhenLocked, Validate, ScriptHooks }
```

File: BondageClub/Scripts/TypedItem.js [hunk 10]
```diff
@@ -103,44 +111,74 @@ function TypedItemCreateTypedItemData(asset,
  * @param {TypedItemData} data - The typed item data for the asset
  * @returns {void} - Nothing
  */
-function TypedItemCreateLoadFunction({ options, functionPrefix, dialog }) {
+function TypedItemCreateLoadFunction({ options, functionPrefix, dialog, scriptHooks}) {
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:114]:
```suggestion
function TypedItemCreateLoadFunction({ options, functionPrefix, dialog, scriptHooks }) {
```

File: BondageClub/Scripts/TypedItem.js [hunk 11]
```diff
@@ -103,44 +111,74 @@ function TypedItemCreateTypedItemData(asset,
...
 	const loadFunctionName = `${functionPrefix}Load`;
-	window[loadFunctionName] = function () {
+	let originalFunction = function () {
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:116]:
```suggestion
	const loadFunction = function () {
```

File: BondageClub/Scripts/TypedItem.js [hunk 12]
```diff
@@ -103,44 +111,74 @@ function TypedItemCreateTypedItemData(asset,
...
 		ExtendedItemLoad(options, dialog.load);
 	};
+	if (scriptHooks && scriptHooks.load) {
+		window[loadFunctionName] = function () {
+			scriptHooks.load(originalFunction);
+		};
+	} else window[loadFunctionName] = originalFunction;
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:123]:
```suggestion
	if (scriptHooks && scriptHooks.load) {
		window[loadFunctionName] = function () {
			scriptHooks.load(loadFunction);
		};
	} else window[loadFunctionName] = loadFunction;
```

File: BondageClub/Scripts/TypedItem.js [hunk 13]
```diff
@@ -103,44 +111,74 @@ function TypedItemCreateTypedItemData(asset,
...
 }
 
 /**
  * Creates an asset's extended item draw function
  * @param {TypedItemData} data - The typed item data for the asset
  * @returns {void} - Nothing
  */
-function TypedItemCreateDrawFunction({ options, functionPrefix, dialog, drawImages }) {
+function TypedItemCreateDrawFunction({ options, functionPrefix, dialog, drawImages, scriptHooks}) {
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:131]:
```suggestion
function TypedItemCreateDrawFunction({ options, functionPrefix, dialog, drawImages, scriptHooks }) {
```

File: BondageClub/Scripts/TypedItem.js [hunk 14]
```diff
@@ -103,44 +111,74 @@ function TypedItemCreateTypedItemData(asset,
...
 	const drawFunctionName = `${functionPrefix}Draw`;
-	window[drawFunctionName] = function () {
+	let originalFunction = function () {
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:133]:
```suggestion
	const drawFunction = function () {
```

File: BondageClub/Scripts/TypedItem.js [hunk 15]
```diff
@@ -103,44 +111,74 @@ function TypedItemCreateTypedItemData(asset,
...
 		ExtendedItemDraw(options, dialog.typePrefix, null, drawImages);
 	};
+	if (scriptHooks && scriptHooks.draw) {
+		window[drawFunctionName] = function () {
+			scriptHooks.draw(originalFunction);
+		};
+	} else window[drawFunctionName] = originalFunction;
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:140]:
```suggestion
	if (scriptHooks && scriptHooks.draw) {
		window[drawFunctionName] = function () {
			scriptHooks.draw(drawFunction);
		};
	} else window[drawFunctionName] = drawFunction;
```

File: BondageClub/Scripts/TypedItem.js [hunk 16]
```diff
@@ -103,44 +111,74 @@ function TypedItemCreateTypedItemData(asset,
...
 }
 
 /**
  * Creates an asset's extended item click function
  * @param {TypedItemData} data - The typed item data for the asset
  * @returns {void} - Nothing
  */
-function TypedItemCreateClickFunction({ options, functionPrefix, drawImages }) {
+ function TypedItemCreateClickFunction({ options, functionPrefix, drawImages, scriptHooks}) {
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:148]:
```suggestion
function TypedItemCreateClickFunction({ options, functionPrefix, drawImages, scriptHooks }) {
```

File: BondageClub/Scripts/TypedItem.js [hunk 17]
```diff
@@ -103,44 +111,74 @@ function TypedItemCreateTypedItemData(asset,
...
 	const clickFunctionName = `${functionPrefix}Click`;
-	window[clickFunctionName] = function () {
+	let originalFunction = function () {
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:150]:
```suggestion
	const clickFunction = function () {
```

File: BondageClub/Scripts/TypedItem.js [hunk 18]
```diff
@@ -103,44 +111,74 @@ function TypedItemCreateTypedItemData(asset,
...
 		ExtendedItemClick(options, null, drawImages);
 	};
+	if (scriptHooks && scriptHooks.click) {
+		window[clickFunctionName] = function () {
+			scriptHooks.click(originalFunction);
+		};
+	} else window[clickFunctionName] = originalFunction;
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:157]:
```suggestion
	if (scriptHooks && scriptHooks.click) {
		window[clickFunctionName] = function () {
			scriptHooks.click(clickFunction);
		};
	} else window[clickFunctionName] = clickFunction;
```

File: BondageClub/Scripts/TypedItem.js [hunk 19]
```diff
@@ -103,44 +111,74 @@ function TypedItemCreateTypedItemData(asset,
...
+		};
+	}
 }
 
 /**
  *
  * @param {TypedItemData} data - The typed item data for the asset
  */
-function TypedItemCreateValidateFunction({ changeWhenLocked, options, functionPrefix, validate }) {
+function TypedItemCreateValidateFunction({ changeWhenLocked, options, functionPrefix, validate, scriptHooks}) {
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:179]:
```suggestion
function TypedItemCreateValidateFunction({ changeWhenLocked, options, functionPrefix, validate, scriptHooks }) {
```

File: BondageClub/Scripts/TypedItem.js [hunk 20]
```diff
@@ -103,44 +111,74 @@ function TypedItemCreateTypedItemData(asset,
...
 	const validateFunctionName = `${functionPrefix}Validate`;
-	window[validateFunctionName] = function (C, item, option, currentOption) {
+	let originalFunction = function (C, item, option, currentOption) {
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:181]:
```suggestion
	const validateFunction = function (C, item, option, currentOption) {
```

File: BondageClub/Scripts/TypedItem.js [hunk 21]
```diff
@@ -154,6 +192,11 @@ function TypedItemCreateValidateFunction({ changeWhenLocked, options, functionPr
 
 		return message;
 	};
+	if (scriptHooks && scriptHooks.validate) {
+		window[validateFunctionName] = function (C, item, option, currentOption) {
+			scriptHooks.validate(originalFunction, C, item, option, currentOption);
+		};
+	} else window[validateFunctionName] = originalFunction;
```
Comment (Jomshir98) [inline BondageClub/Scripts/TypedItem.js:199]:
```suggestion
	if (scriptHooks && scriptHooks.validate) {
		window[validateFunctionName] = function (C, item, option, currentOption) {
			scriptHooks.validate(validateFunction, C, item, option, currentOption);
		};
	} else window[validateFunctionName] = validateFunction;
```

File: BondageClub/Scripts/Typedef.d.ts [hunk 22]
```diff
@@ -823,6 +823,16 @@ interface ModularItemConfig {
...
 	ChangeWhenLocked?: boolean;
+	/**
+	 * A recond containing functions that are run on load, click, draw, exit, and validate, with the original archetype function
+	 * and parameters passed on to them. If undefined, these are ignored
+	 */
+	ScriptHooks?: {
+		Load? : (OriginalFunction : () => void) => void,
+		Click? : (OriginalFunction : () => void) => void,
+		Draw? : (OriginalFunction : () => void) => void,
+		Exit? : () => void,
```
Comment (Jomshir98) [inline BondageClub/Scripts/Typedef.d.ts:834]:
```suggestion
		Load?: (next: () => void) => void,
		Click?: (next: () => void) => void,
		Draw?: (next: () => void) => void,
		Exit?: () => void,
```

File: BondageClub/Scripts/Typedef.d.ts [hunk 23]
```diff
@@ -914,6 +924,17 @@ interface ModularItemData {
...
+	/**
+	 * A recond containing functions that are run on load, click, draw, exit, and validate, with the original archetype function
+	 * and parameters passed on to them. If undefined, these are ignored
+	 */
+	scriptHooks?: {
+		load? : (OriginalFunction : () => void) => void,
+		click? : (OriginalFunction : () => void) => void,
+		draw? : (OriginalFunction : () => void) => void,
+		exit? : () => void,
+
```
Comment (Jomshir98) [inline BondageClub/Scripts/Typedef.d.ts:936]:
```suggestion
		load?: (next: () => void) => void,
		click?: (next: () => void) => void,
		draw?: (next: () => void) => void,
		exit?: () => void,
```

File: BondageClub/Scripts/Typedef.d.ts [hunk 24]
```diff
@@ -964,6 +985,17 @@ interface TypedItemConfig {
...
+	/**
+	 * A recond containing functions that are run on load, click, draw, exit, and validate, with the original archetype function
+	 * and parameters passed on to them. If undefined, these are ignored
+	 */
+	ScriptHooks?: {
+		Load? : (OriginalFunction : () => void) => void,
+		Click? : (OriginalFunction : () => void) => void,
+		Draw? : (OriginalFunction : () => void) => void,
+		Exit? : () => void,
+		Validate? : (OriginalFunction : () => string, C : Character, Item : Item, Option : ExtendedItemOption, CurrentOption : ExtendedItemOption) => string,
```
Comment (Jomshir98) [inline BondageClub/Scripts/Typedef.d.ts:997]:
```suggestion
		Load?: (next: () => void) => void,
		Click?: (next: () => void) => void,
		Draw?: (next: () => void) => void,
		Exit?: () => void,
		Validate? : (next: () => string, C: Character, Item: Item, Option: ExtendedItemOption, CurrentOption: ExtendedItemOption) => string,
```

File: BondageClub/Scripts/Typedef.d.ts [hunk 25]
```diff
@@ -1043,6 +1075,17 @@ interface TypedItemData {
...
+	/**
+	 * A recond containing functions that are run on load, click, draw, exit, and validate, with the original archetype function
+	 * and parameters passed on to them. If undefined, these are ignored
+	 */
+	scriptHooks?: {
+		load? : (OriginalFunction : () => void) => void,
+		click? : (OriginalFunction : () => void) => void,
+		draw? : (OriginalFunction : () => void) => void,
+		exit? : () => void,
+		validate? : (OriginalFunction : () => string, C : Character, Item : Item, Option : ExtendedItemOption, CurrentOption : ExtendedItemOption) => string,
```
Comment (Jomshir98) [inline BondageClub/Scripts/Typedef.d.ts:1087]:
```suggestion
		load?: (next: () => void) => void,
		click?: (next: () => void) => void,
		draw?: (next: () => void) => void,
		exit?: () => void,
		validate?: (next: () => string, C: Character, Item: Item, Option: ExtendedItemOption, CurrentOption: ExtendedItemOption) => string,
```

**Comment by Ben987** [comment]
This is HUGE!  Lots of work in there, I'm curious to try that.  Thanks a lot :)

**Pull request merged**